### PR TITLE
Add new widget area below the dockpanel

### PR DESCRIPTION
**Pull request by fcollonval**
## References

This aims to address #9735 by adding a tabpanel below the main dockpanel.

TODO and questions:
- [x] Add new panel to layout restorer
  - [x] SplitPanel is missing a signal when layout is changed to be stored ~~=> needs a change upstream in lumino.~~ Use message hook on the new vertical split panel.
- [x] Add focused element on the new tabpanel
- [x] doc
- [x] tests
- What about the integration in simple mode? Should that new panel be hidden?
- Quid on accessibility navigation?

## Code changes

The dockpanel is now wrapped in a vertical split panel to insert a tabpanel below it.

The log console opens now in that tabpanel.

## User-facing changes

The idea is to separate the concern of documents and other kind of panels (for now log console).

The widget in the dock panel cannot be moved in the new tabpanel and vice-versa.

The tabpanel is hidden if empty.

In the demo below the [dev js console](https://github.com/QuantStack/jupyterlab-js-logs) was installed too to get two widgets in the new panel.
downpanel

## Backwards-incompatible changes

N/A

### CODE REVIEW

**Comment by blink1073** [comment]
cc @ellisonbg @isabela-pf

File: packages/application/src/shell.ts [hunk 1]
```diff
@@ -87,7 +89,8 @@ export namespace ILabShell {
     | 'menu'
     | 'left'
     | 'right'
-    | 'bottom';
+    | 'bottom'
+    | 'down';
```
Comment (jtpio) [inline packages/application/src/shell.ts:93]: For extension authors, we'll probably want to document this in the extension developer guide too: https://jupyterlab.readthedocs.io/en/stable/extension/extension_points.html#jupyter-front-end-shell

**Comment by fcollonval** [comment]
Regarding the simple mode, I think the current design has a flaw for the console and show contextual help panels as they cannot be seen on the same tab as the notebook (see example below with the console):

log_simple_mode

Adding this panel down would allow to display them on the same tab than the notebook.

This would also make sense for some extension like the diagnostic panel of JupyterLab LSP (cc @bollwyvl @krassowski)

**Comment by krassowski** [comment]
> This would also make sense for some extension like the diagnostic panel of JupyterLab LSP

Yes, indeed. We could make it a setting to open the diagnostics panel there (not a default though because some users use it on the side and the inability to drag it would be a regression). If things cannot be dragged into this space (nor out of there), would it make sense to add a different CSS style to the bottom area?

Also, It would make sense for me to be able to drag things between sidebars and the bottom panel. I know it is more difficult implementation-wise, but as user I would enjoy it.

File: packages/application/src/shell.ts (hunk 1 above)
Comment (fcollonval) [reply packages/application/src/shell.ts:98]: Thanks for the suggestion @jtpio. I updated the doc.

**Comment by fcollonval** [comment]
Here is a comparison for the help panel between the classical notebook, the current JLab and this proposal:

**Classical notebook**

notebook_help

**This PR**

pr_jlab_help

**Current JupyterLab**

current_jlab_help

**Comment by fcollonval** [comment]
> We could make it a setting to open the diagnostics panel there (not a default though because some users use it on the side and the inability to drag it would be a regression). 

> If things cannot be dragged into this space (nor out of there), would it make sense to add a different CSS style to the bottom area?

As UX is not my strong side, I let @ellisonbg and @isabela-pf suggest direction to follow.

> Also, It would make sense for me to be able to drag things between sidebars and the bottom panel. I know it is more difficult implementation-wise, but as user I would enjoy it.

It sounds a good addition indeed. But it that case, I would rather not implement the possibility to move it to the main area (as you suggested for the LSP diagnostic panel).

**Comment by fcollonval** [comment]
This is ready for review. Tab motion by drag-and-drop is left aside for another PR.

**Comment by krassowski** [comment]
I am concerned about the inability to move the Contextual Help to the side. I use it this way a lot, and I believe that other users with monitors which are wider rather than taller use it this way too. This is a substantial restrictive change, and I think that it might annoy users.

But maybe I am doing something wrong? Screencast from binder test below:

contextual-to-side

**Comment by isabela-pf** [comment]
Thanks for making this PR and all the gifs comparing behaviors! Reading the issue this is based on definitely helped me understand what the PR does, but I'm still unclear why this needs to be it's own new widget area that doesn't allow tabs to be moved to other areas. Like @krassowski pointed out, there's some ways I'd anticipate this limit creating confusion or frustration, especially since it (logically) looks like all the pre-existing widget areas.

My ideal interaction would be to have the log default open in this new area, but still have it available to move to other widget areas. I think this would match current JLab interactions and user expectations while still providing a new place to group all the different tools listed in the issue. 

I'm hesitant to involve the settings for two reasons: 
1. It would be easy to miss that something like that is configurable
2. The current settings experience can be overwhelming

Thank you for your patience!

**Comment by fcollonval** [comment]
Thanks for the feedback @isabela-pf 

> I'm still unclear why this needs to be it's own new widget area that doesn't allow tabs to be moved to other areas. Like @krassowski pointed out, there's some ways I'd anticipate this limit creating confusion or frustration, especially since it (logically) looks like all the pre-existing widget areas.

There are multiple actors pushing to a IDE style interface. In that interface style, you get an area down the window (e.g. Visual Studio/Code, PyCharm, Eclipse, Qt Creator) for logs, terminals,... (mainly things that are not the main work documents but that require larger width compared to a sidebar).

> My ideal interaction would be to have the log default open in this new area, but still have it available to move to other widget areas. I think this would match current JLab interactions and user expectations while still providing a new place to group all the different tools listed in the issue.

This is gonna be tricky if we don't allow all widgets to move from that area to the main area.

In light of the discussion, what about introducing a new interface mode:
- Simple => The new area will be activated for logs and contextual help
- Default => No changes
- IDE-like => The new area will be activated for logs and contextual help

The tab bar style of the new area will be changed to highlight this is a separated area.

In a follow-up PR, I'll implement the drag-and-drop between sidebars and this new panel.

**Comment by goanpeca** [comment]
Now that I read more on this PR, it feels like this is trying to solve a larger JLab problem with an intermediate solution. I have opened this https://github.com/jupyterlab/jupyterlab/issues/8960 and it seems like having a generic layout solution in JLab, instead of new areas would provide a general solution for most cases?

**Comment by fcollonval** [comment]
> Now that I read more on this PR, it feels like this is trying to solve a larger JLab problem with an intermediate solution. I have opened this #8960 and it seems like having a generic layout solution in JLab, instead of new areas would provide a general solution for most cases?

This is hard to say. Trying to summarize your suggestion: this implies that all tabs of a certain types will get in the same stack panel with those panels gathered in a dockpanel; with maybe a special case for the document panel that can allow split of documents?

**Comment by goanpeca** [comment]
>This is hard to say. Trying to summarize your suggestion: this implies that all tabs of a certain types will get into the same stack panel with those panels gathered in a dockpanel; with maybe a special case for the document panel that can allow split of documents?

Something like that. Not saying that it should block this PR, I do see the usefulness.

A long term solution would probably be more in the way of using "predefined layouts". It may be that predefined layouts actually implies creating new areas for every place in a given grid (2x2, 3x2, 1x2 etc...) and in that case it makes sense that things cannot move between areas.

**Comment by bollwyvl** [comment]
If there _was_ a bottom drawer, i'd see that being full-width, and collapsing the status bar into it. This would give a convenient, consistent way to provide more context vs popups.

If an item in the status bar item offered an expanded view, and it was selected, the drawer would slide up. Alas, tab bars are a magic snowflake lumino thing :scream: but maybe maybe a FatTabBarPanel could be created for this purpose where the tabs were actual widgets.

Further, i'd rather see the all the areas become equivalent drop targets, rather than _further_ constraining it. This would make it easy to drag a main to a sidebar (removing the need for special-case extensions like Sidecar). The main area would still only be controlled by the "simple" switch. Indeed, the top bar could also receive this treatment, with the huge document name banner becoming collapsible, such that one could actually get down to a mere 48px of "wasted" space.

**Comment by fcollonval** [comment]
> If an item in the status bar item offered an expanded view, and it was selected, the drawer would slide up.

I'm not sure this is a good idea. You will have elements purely informative mixed with panel opener. This may confuse users. I looked at some IDE; they all have separated the status bar and the tab bar opening down panels:

PyCharm
image

Wing Python IDE
image

Spyder
image

VSCode
image

**Comment by isabela-pf** [comment]
> This is gonna be tricky if we don't allow all widgets to move from that area to the main area.

@fcollonval I agree! At the moment, I don't see a reason for not having all widgets being able to move in or out of this area. I understand the idea is to have a new widget area to provide a familiar experience from other IDEs and to group similar outputs, but since all other widget areas allow for the user to personalize I think it would be unpleasant to have a new widget be the only widget area in the interface to break the pattern. It still makes sense to me to change these outputs to default to this new area, but I also think it makes sense to be able to move them into any other widget area. 

I think this is what @bollwyvl is saying here
> Further, i'd rather see the all the areas become equivalent drop targets, rather than further constraining it.
so I think it's not just me thinking this. Does that make more sense?

**Comment by fcollonval** [comment]
Thanks for all comments. The way forward I see is the following:
- [ ] Make the new area full width even if a side bar is opened
- [ ] Style the tabbar to be consistent with the sidebar. So it highlights the difference
- [ ] Allow drag and drop between all areas => in a follow-up PR

**Comment by fcollonval** [comment]
This is ready for review.

The area is now full width and the tabbar is styled as the sidebars.
For now only the log console open in that area. The inspector default anchorage will be changed in the follow-up PR introducing panel drag-and-drop between sidebars and this new area.

Screenshots:

| Theme | Light | Dark |
| --- | --- | --- |
| single | light_single | dark_single |
| multiple | light_multiple | dark_multiple |

**Review by blink1073 — APPROVED** [review]
Looks good, thank you!

File: packages/application/src/frontend.ts [hunk 2]
```diff
@@ -283,6 +283,11 @@ export namespace JupyterFrontEnd {
      */
     readonly currentWidget: Widget | null;
 
+    /**
+     * The user interface mode.
+     */
+    readonly mode: DockPanel.Mode;
```
Comment (jtpio) [inline packages/application/src/frontend.ts:289]: Was this addition intentional?

This means that starting with `^3.1.0-alpha.12`, custom apps implementing a `JupyterFrontEnd.IShell` now need to also implement `mode`.

Is this something we want to keep? It looks like the mode is not used in the context of this PR.

Comment (jtpio) [reply packages/application/src/frontend.ts:289]: This was noticed when updating RetroLab to use the alpha 12 packages: https://github.com/jupyterlab/retrolab/pull/154/commits/7756e96ef86c7211c24a17b30e0b912c4d9171cf
